File: packages/intellij/src/main/kotlin/com/spek/intellij/core/OpenspecCli.kt
package com.spek.intellij.core

import com.intellij.openapi.application.ApplicationManager
import java.io.File
import java.util.concurrent.Callable
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ExecutionException
import java.util.concurrent.FutureTask
import java.util.concurrent.TimeUnit

/**
 * Talking to the `openspec` CLI, and remembering what it said.
 *
 * The Kotlin counterpart of `packages/core/src/openspec-cli.ts`, and it exists for the same reason
 * that module does: both of these were written twice before — once in [SchemaOrder] and again in
 * [SchemaCatalog] — and the duplication was not free. On the TypeScript side the cache's original
 * "remember failures forever" bug had to be found once and then hand-copied as a fix into the second
 * copy. The subprocess safeguards and the cache policy are each stated once, here.
 */
object OpenspecCli {
    /**
     * How long the CLI gets before it is killed.
     *
     * Load-bearing against [TtlCache]'s default TTL, which must stay >= this — see the note there.
     */
    const val TIMEOUT_SECONDS = 10L

    /**
     * The raw outcome of running the CLI — deliberately *not* a verdict.
     *
     * Interpretation differs by caller and belongs to them: [SchemaOrder] collapses every failure to
     * null, while [SchemaCatalog] needs a failure taxonomy *and* stdout even on a non-zero exit,
     * because `openspec schema which <unknown>` answers correctly and exits 1.
     */
    sealed interface Outcome {
        data class Completed(val exitCode: Int, val stdout: String) : Outcome
        data object TimedOut : Outcome
        data object StartFailed : Outcome
    }

    /**
     * Reasons a second read could plausibly come out differently — the ones a running host repairs
     * by itself: a binary not on `PATH` *yet*, or load easing before the next call.
     *
     * The other two are not. A non-zero exit with no body, or output that cannot be parsed, is the
     * *installed CLI* answering — a version too old for a command it still marks experimental, a
     * wrapper printing a banner — and a read a second later finds it identical. Retrying those on
     * every read spawns a process to be told the same thing, on views that re-read whenever the
     * watched tree changes.
     *
     * The mirror of `isTransient` in `openspec-cli.ts`, and stated once here for the same reason.
     */
    fun isTransient(reason: SchemaDegradedReason): Boolean =
        reason == SchemaDegradedReason.CLI_UNAVAILABLE || reason == SchemaDegradedReason.CLI_TIMEOUT

    /**
     * Run `openspec <args>` in [cwd] and return its exit code and stdout.
     *
     * Four safeguards, all load-bearing:
     * 1. stderr is DISCARDed rather than merged — the schema commands print an experimental-command
     *    notice there, which would corrupt stdout's JSON, and an undrained stderr pipe can fill and
     *    deadlock the child.
     * 2. stdout is drained on an IDE pooled thread, so a full pipe cannot block the child. The IDE
     *    pool is cached-type and suited to blocking IO; `commonPool` has parallelism = cores-1, where
     *    a blocking read can starve the reader task on a low-core machine and make a healthy CLI look
     *    like a timeout.
     * 3. `waitFor` has a hard timeout and `destroyForcibly` follows it — cancelling the reader future
     *    cannot interrupt a blocking read; closing the child's stdout is what releases it.
     * 4. the reader's own `get` is bounded too.
     *
     * **Callers must validate every interpolated argument first.** On Windows, ProcessBuilder's argv
     * is re-parsed by cmd.exe for `openspec.cmd` (BatBadBut / CVE-2024-27980) and, unlike Node's
     * cross-spawn on the TypeScript side, it does not escape for us — so an allowlist at the call
     * site is a security boundary here, not a tidiness rule.
     */
    fun run(args: List<String>, cwd: String, timeoutSeconds: Long = TIMEOUT_SECONDS): Outcome {
        return try {
            val bin = if (System.getProperty("os.name").orEmpty().lowercase().contains("win")) {
                "openspec.cmd"
            } else {
                "openspec"
            }
            val proc = ProcessBuilder(listOf(bin) + args)
                .directory(File(cwd))
                .redirectError(ProcessBuilder.Redirect.DISCARD)
                .start()
            val reader = ApplicationManager.getApplication().executeOnPooledThread(
                Callable { proc.inputStream.bufferedReader().use { it.readText() } },
            )
            if (!proc.waitFor(timeoutSeconds, TimeUnit.SECONDS)) {
                proc.destroyForcibly()
                reader.cancel(true)
                return Outcome.TimedOut
            }
            Outcome.Completed(proc.exitValue(), reader.get(2, TimeUnit.SECONDS))
        } catch (_: Exception) {
            // openspec not installed → IOException from start(); anything else is equally unusable.
            Outcome.StartFailed
        }
    }

    /**
     * Run [tasks] concurrently on IDE pooled threads and collect the results in order.
     *
     * Falls back to running them in sequence when there is no Application — plain unit tests have no
     * IDE, and this is an optimisation, not a requirement. Failing there would make the tests depend
     * on a platform they otherwise do not need.
     */
    fun <T> inParallel(tasks: List<() -> T>): List<T> {
        val app = ApplicationManager.getApplication()
        if (tasks.size <= 1 || app == null) return tasks.map { it() }
        return tasks.map { task -> app.executeOnPooledThread(Callable { task() }) }.map { it.get() }
    }
}

/**
 * A TTL- and size-capped cache holding one **computation** per key.
 *
 * A TTL because caching a failure forever meant a reader who installed the CLI after first load
 * never got data without restarting the IDE; it must stay >= [OpenspecCli.TIMEOUT_SECONDS] so an
 * in-flight call is never judged stale. A size cap because these are application-level singletons
 * shared across every project window. (`cli-budget.ts` states the same bounds.) The TTL is the
 * bound on a remembered *answer*; a failure worth retrying is not held for it at all — see
 * [Outcome].
 *
 * `ConcurrentHashMap` because handlers arrive on Netty threads. CHM forbids null values, so entries
 * wrap the result — [V] may be nullable, since "computed, and there is none" is an answer like any
 * other. CHM has no insertion order, so the cap evicts an arbitrary entry rather than the oldest.
 *
 * The entry holds a [FutureTask] so a caller arriving mid-flight joins the run rather than starting
 * a second one — storing the value instead made two concurrent callers each run the whole CLI
 * enumeration for one answer. `putIfAbsent` + `run()`, **not** `computeIfAbsent`, which would hold
 * the bin lock across a subprocess.
 *
 * What is remembered is decided by [compute], which returns an [Outcome] saying so — see
 * `openspec-cli.ts`'s `CliOutcome`, of which this is the mirror. The cache cannot judge it from the
 * value, because by then the value has usually lost the distinction: a null schema order is both
 * "the CLI answered and there is no order" and "the CLI could not be reached".
 */
class TtlCache<K : Any, V>(
    private val ttlMs: Long = 30_000L,
    private val maxSize: Int = 256,
) {
    /**
     * A result, plus whether it is worth remembering.
     *
     * Covariant because it only ever hands [value] out: without that, `failed(null)` and
     * `answered(order)` in one `when` infer as unrelated types and every call site has to spell the
     * type parameter out.
     */
    class Outcome<out V> private constructor(val value: V, val remember: Boolean) {
        companion object {
            /** An answer: kept for the TTL, like any successful read. */
            fun <V> answered(value: V): Outcome<V> = Outcome(value, true)

            /** A failure the next read could find gone: returned to the caller, but not kept. */
            fun <V> failed(value: V): Outcome<V> = Outcome(value, false)
        }
    }

    /** A held value, boxed so an absent entry stays distinguishable from an entry holding null. */
    class Held<out V>(val value: V)

    private class Entry<V>(val at: Long, val task: FutureTask<Outcome<V>>)

    private val map = ConcurrentHashMap<K, Entry<V>>()

    /**
     * What [key] currently holds, or null — **without installing anything**.
     *
     * [getOrCompute] cannot answer this: asking it is asking it to compute. A caller with something
     * cheaper than a CLI call to do when there is no current entry has to find that out without
     * creating one — see [SchemaOrder], which replays a settled failure only when the bucket has no
     * answer to serve. Joins an in-flight run like any hit; an expired entry reads as absent and is
     * left for [getOrCompute] to replace, so a peek cannot race out a run installed since.
     */
    fun peek(key: K): Held<V>? {
        val existing = map[key] ?: return null
        if (System.currentTimeMillis() - existing.at > ttlMs) return null
        return Held(existing.task.awaitValue().value)
    }

    fun getOrCompute(key: K, compute: () -> Outcome<V>): V {
        while (true) {
            val existing = map[key]
            if (existing != null) {
                // A joiner holds the entry itself, so a failure removed from the map after it
                // resolves is still delivered here — sharing the run does not depend on keeping it.
                if (System.currentTimeMillis() - existing.at <= ttlMs) return existing.task.awaitValue().value
                // Two-arg remove, so a fresher entry installed since the read is not discarded.
                map.remove(key, existing)
            }

            val task = FutureTask(compute)
            val entry = Entry(System.currentTimeMillis(), task)
            val raced = map.putIfAbsent(key, entry)
            // Someone else got there first; loop so their entry is freshness-checked like any hit.
            if (raced != null) continue
            if (map.size > maxSize) {
                // Never the entry just installed, or a burst evicts the run everyone is waiting on.
                map.keys.firstOrNull { it != key }?.let { map.remove(it) }
            }
            task.run()
            // The thread that installed the entry is the only one that runs it, so it is the one
            // that drops it. `finally` rather than a check on the result: a compute that threw
            // produces no outcome to inspect, and a remembered exception is the same bug in another
            // shape — every caller in the window handed a failure whose cause may be gone.
            var remember = false
            try {
                val outcome = task.awaitValue()
                remember = outcome.remember
                return outcome.value
            } finally {
                if (!remember) map.remove(key, entry)
            }
        }
    }

    fun clear() = map.clear()

    /** Unwrapped, so a shared run reports the same exception an unshared one would. */
    private fun FutureTask<Outcome<V>>.awaitValue(): Outcome<V> =
        try {
            get()
        } catch (e: ExecutionException) {
            throw e.cause ?: e
        }
}

/**
 * A TTL- and size-capped set of keys: the record that something happened, carrying no value.
 *
 * Beside [TtlCache] rather than in a module of its own, and taking the same bounds, because the one
 * thing it must never do is age differently from the cache it complements — [SchemaOrder] keeps a
 * cache of answers and a set of settled changes, and a mark outliving an answer is a mark denying a
 * consultation the cache would already have re-run.
 *
 * `ConcurrentHashMap` for the reason [TtlCache] uses one: handlers arrive on Netty threads. CHM has
 * no insertion order, so the cap evicts an arbitrary key rather than the oldest — a bound, not a
 * policy.
 */
class TtlMarks<K : Any>(
    private val ttlMs: Long = 30_000L,
    private val maxSize: Int = 256,
) {
    private val map = ConcurrentHashMap<K, Long>()

    fun isMarked(key: K): Boolean {
        val at = map[key] ?: return false
        if (System.currentTimeMillis() - at > ttlMs) {
            map.remove(key, at)
            return false
        }
        return true
    }

    fun mark(key: K) {
        if (map.size >= maxSize) map.keys.firstOrNull { it != key }?.let { map.remove(it) }
        map[key] = System.currentTimeMillis()
    }

    fun clear() = map.clear()
}

File: packages/core/src/openspec-cli.ts
import spawn from "cross-spawn";
import { CLI_CACHE_TTL_MS, CLI_TIMEOUT_MS } from "./cli-budget.js";
import type { SchemaDegradedReason } from "./types.js";

/**
 * Talking to the `openspec` CLI, and remembering what it said.
 *
 * Both of the things here were written twice before this module existed — once in `schema-order.ts`
 * and again in `schemas.ts` — and the duplication was not free: the cache's original "remember
 * failures forever" bug had to be found once and then hand-copied as a fix into the second copy.
 * The 10s timeout, the spawn options, the 30s/256-entry cache policy, and which failures are worth
 * remembering at all are each stated once, here.
 */

// Stryker disable all: thin integration layer over a child process. The parsing it feeds is
// unit-tested; this only spawns, times out, and classifies failure. The cache below is the
// exception — it holds policy, and `openspec-cli.test.ts` covers it.

// Defined in cli-budget.ts so browser bundles can read them without pulling in child_process.
export { CLI_TIMEOUT_MS } from "./cli-budget.js";

/**
 * Every failure mode gets its own reason so surfaces can word them differently — "the CLI is not
 * installed" is fixable by the reader; "the CLI returned something we could not read" is not the
 * same thing at all. Callers that do not care collapse the whole `!ok` side to a single fallback.
 *
 * A failure may still carry `json`. The CLI exits non-zero for *domain* answers as well as for
 * breakage — `schema which <unknown>` prints a perfectly good `{"error": "Schema 'x' not found"}`
 * and exits 1 — so discarding stdout on a non-zero exit loses the answer and reports a working CLI
 * as broken. Parsed stdout is kept; deciding what it means belongs to the caller.
 */
export type CliResult =
  | { ok: true; json: unknown }
  | { ok: false; reason: SchemaDegradedReason; json?: unknown };

/**
 * Reasons a second read could plausibly come out differently — the ones a running host repairs by
 * itself. `cli-unavailable` is a binary not on `PATH` *yet* (a desktop app resolving the user's
 * shell `PATH` at startup is the reported case, issue #46); `cli-timeout` is load easing.
 *
 * The other two are not: a non-zero exit with no body, or output that cannot be parsed, is the
 * *installed CLI* answering — a version too old for a command it still marks experimental, a wrapper
 * printing a banner — and a read a second later finds it identical. Retrying those on every read
 * buys nothing and costs a full process start each time (~0.65-1.3s here, against ~5ms for a missing
 * binary), on views that re-read whenever the watched tree changes.
 *
 * One rule rather than a list per caller, and it reads `SchemaDegradedReason`, so a reason added
 * later has to be placed on one side of it.
 */
export function isTransient(reason: SchemaDegradedReason): boolean {
  return reason === "cli-unavailable" || reason === "cli-timeout";
}

/**
 * The CLI's own error message, when a failed run still produced a structured `{ error: string }`
 * body. Non-null means the CLI **ran and answered** — the request failed, not the tool.
 */
export function cliErrorMessage(json: unknown): string | null {
  if (!json || typeof json !== "object") return null;
  const message = (json as Record<string, unknown>).error;
  return typeof message === "string" ? message : null;
}

/** Run `openspec <args>` and parse its stdout as JSON. */
function spawnOpenspecJson(args: string[], cwd: string): Promise<CliResult> {
  // argv array, not a shell string: each argument is structurally its own token, so no input can
  // become shell syntax. cross-spawn resolves openspec.cmd on Windows without a shell (avoiding
  // DEP0190 / CVE-2024-27980).
  return new Promise<CliResult>((resolve) => {
    let out = "";
    let settled = false;
    let timer: ReturnType<typeof setTimeout>;
    const finish = (r: CliResult) => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      resolve(r);
    };
    let child;
    try {
      child = spawn("openspec", args, {
        cwd,
        // stderr ignored deliberately rather than merged: the schema commands print an
        // experimental-command notice there, and merging it would corrupt stdout's JSON.
        stdio: ["ignore", "pipe", "ignore"],
        windowsHide: true,
      });
    } catch {
      finish({ ok: false, reason: "cli-unavailable" });
      return;
    }
    timer = setTimeout(() => {
      child.kill();
      finish({ ok: false, reason: "cli-timeout" });
    }, CLI_TIMEOUT_MS);
    child.stdout?.setEncoding("utf-8");
    child.stdout?.on("data", (d: string) => {
      out += d;
    });
    child.on("error", () => finish({ ok: false, reason: "cli-unavailable" })); // ENOENT
    child.on("close", (code: number | null) => {
      let json: unknown;
      let parsed = false;
      try {
        json = JSON.parse(out);
        parsed = true;
      } catch {
        // leave unparsed
      }
      if (code !== 0) {
        // stdout is still forwarded when it parsed — see the note on CliResult.
        finish(parsed ? { ok: false, reason: "cli-failed", json } : { ok: false, reason: "cli-failed" });
        return;
      }
      finish(parsed ? { ok: true, json } : { ok: false, reason: "cli-unparsable" });
    });
  });
}

/** Injectable CLI runner, so tests drive a stub instead of the real binary. */
export type OpenspecRunner = (args: string[], cwd: string) => Promise<CliResult>;

let runner: OpenspecRunner = spawnOpenspecJson;

/** Replace the CLI runner (tests only). Returns the previous one so callers can restore it. */
export function setOpenspecRunner(next: OpenspecRunner): OpenspecRunner {
  const prev = runner;
  runner = next;
  return prev;
}

/** Run the CLI through whichever runner is installed. */
export function runOpenspec(args: string[], cwd: string): Promise<CliResult> {
  return runner(args, cwd);
}

// ---------------------------------------------------------------------------
// Caching
// ---------------------------------------------------------------------------

/**
 * TTL and its floor are stated in cli-budget.ts; the size cap depends on nothing and is stated here.
 *
 * Both are exported because a caller may need a second, differently-keyed store beside the one it
 * memoises through — `schema-order.ts` keeps one — and such a store has to age and be bounded on the
 * same terms. A second copy of either number is a second thing to keep in step, which is the whole
 * reason this module exists.
 */
export const CACHE_TTL_MS = CLI_CACHE_TTL_MS;
export const CACHE_MAX = 256;

export interface CacheEntry<T> {
  at: number;
  promise: Promise<T>;
}

/**
 * A result, plus whether it is worth remembering.
 *
 * The cache cannot work this out from the value, because the value has usually already lost it: a
 * null schema order means both "the CLI answered and there is no order" and "the CLI could not be
 * reached", and a definition read reports "no such schema" both for a name the CLI refused and for a
 * `schema.yaml` it could not open. So whoever still holds the two apart says which this is, and says
 * it *there* — a predicate taking the value could only re-derive a judgement that has already been
 * thrown away.
 */
export interface CliOutcome<T> {
  value: T;
  remember: boolean;
}

/** An answer: cached for the bounded lifetime, like any successful read. */
export function answered<T>(value: T): CliOutcome<T> {
  return { value, remember: true };
}

/** A failure the next read could find gone: returned to the caller, but not remembered. */
export function failed<T>(value: T): CliOutcome<T> {
  return { value, remember: false };
}

/**
 * Memoise `compute` under `key`, holding the **Promise** rather than the resolved value — so
 * concurrent callers arriving mid-flight share one run instead of starting a second.
 *
 * `compute` must classify its own result. There is deliberately no default: every call site that
 * existed when this was added had got it wrong (a failure was remembered exactly as long as an
 * answer, so one unreachable CLI meant half a minute of stale "unavailable" — issue #46), and a
 * default is how the next one would get it wrong without saying anything.
 *
 * A failure is dropped **after** it resolves, not skipped: while it is in flight it is a run like
 * any other, and joiners must share it. Not remembering a failure must not become not deduping one,
 * or a host with no `openspec` on `PATH` spawns a process per concurrent reader.
 */
/**
 * What `key` currently holds, or `undefined` — **without installing anything**.
 *
 * `ttlCached` cannot answer this: asking it is asking it to compute. A caller that has something
 * cheaper than a CLI call to do when there is no current entry needs to find that out without
 * creating one, and creating one is exactly what would make the cheaper path wrong (see
 * `schema-order.ts`, which replays a settled failure only when the bucket has no answer to serve).
 *
 * An expired entry reads as absent and is left in place: dropping it is `ttlCached`'s to do, on the
 * call that replaces it, and a peek that deleted could race a run installed since.
 */
export function peekCached<T>(store: Map<string, CacheEntry<T>>, key: string): Promise<T> | undefined {
  const hit = store.get(key);
  if (!hit || Date.now() - hit.at > CACHE_TTL_MS) return undefined;
  return hit.promise;
}

export function ttlCached<T>(
  store: Map<string, CacheEntry<T>>,
  key: string,
  compute: () => Promise<CliOutcome<T>>,
): Promise<T> {
  const hit = store.get(key);
  if (hit && Date.now() - hit.at <= CACHE_TTL_MS) return hit.promise;
  if (hit) store.delete(key);

  const outcome = compute();
  const promise = outcome.then((o) => o.value);
  if (store.size >= CACHE_MAX) {
    const oldest = store.keys().next().value; // Map keeps insertion order
    if (oldest !== undefined) store.delete(oldest);
  }
  const entry: CacheEntry<T> = { at: Date.now(), promise };
  store.set(key, entry);

  // Identity-checked, so an entry installed since — by a caller that arrived after this one was
  // dropped, or after the TTL — is not deleted by this run's late verdict.
  const forget = (): void => {
    if (store.get(key) === entry) store.delete(key);
  };
  // A rejection is forgotten too. A thrown error held for the full lifetime hands every caller in
  // the window the same failure, which is this bug with an exception in place of a value.
  outcome.then((o) => {
    if (!o.remember) forget();
  }, forget);
  return promise;
}

// Stryker restore all
